File: backend/app/schemas/extraction_analytics.py
"""Schemas for the admin Extraction Analytics page (`/settings/extractions`).

Read-only observability over what each external enrichment provider
(Apollo, Hunter, Snov, PDL, site-crawler, theHarvester, and the website
resolvers) extracted, grouped by firm. Three response shapes:

- ``ProviderSummaryResponse`` — the KPI strip + per-provider rollup.
- ``FirmExtractionListResponse`` — paginated firm-grouped list, each firm
  carrying per-provider counts.
- ``FirmExtractionDetailResponse`` — drill-down to the actual extracted
  values for one firm, each tagged with the provider that produced it.

Kept in its own namespace (not ``schemas/stats.py``) so this admin-only
surface can widen independently of the user-facing dashboard stats.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel

FirmType = Literal["broker_dealer", "advisor", "institutional_investor"]
Surface = Literal["contact", "discovered_email", "website"]
ValueKind = Literal["email", "phone", "linkedin", "website"]


class ProviderCount(BaseModel):
    """One provider's rollup within a surface (contacts / emails / websites).

    ``value_count`` is rows extracted (contacts won, emails discovered, or
    websites resolved); ``firm_count`` is the distinct firms touched.
    """

    provider: str
    label: str
    surface: Surface
    firm_count: int
    value_count: int


class ProviderSummaryTotals(BaseModel):
    total_contacts_attributed: int
    # Contacts whose ``discovery_source`` is NULL (legacy rows that predate
    # provider attribution). Surfaced so the count isn't silently dropped.
    total_contacts_unattributed: int
    total_discovered_emails: int
    total_enriched_emails: int
    total_firms_with_website: int


class ProviderSummaryResponse(BaseModel):
    contacts: list[ProviderCount]
    discovered_emails: list[ProviderCount]
    websites: list[ProviderCount]
    totals: ProviderSummaryTotals


class FirmProviderCount(BaseModel):
    provider: str
    label: str
    count: int


class FirmExtractionRow(BaseModel):
    """One firm in the grouped list, with per-provider extraction counts."""

    firm_type: FirmType
    firm_id: int
    name: str
    crd_number: str | None
    website_source: str | None
    contact_total: int
    discovered_email_total: int
    # Non-zero provider counts for this firm, highest first. Counts reflect
    # the winning provider per contact (``discovery_source``); expand the
    # firm for per-value provenance.
    provider_counts: list[FirmProviderCount]
    last_enriched_at: datetime | None


class FirmExtractionListResponse(BaseModel):
    items: list[FirmExtractionRow]
    total: int


class ExtractedValue(BaseModel):
    """A single extracted datum and the provider that produced it."""

    kind: ValueKind
    value: str
    provider: str
    label: str
    contact_id: int | None
    contact_name: str | None
    contact_title: str | None
    confidence: float | None
    source_table: Literal["contact", "discovered_email", "firm"]


class FirmExtractionDetailResponse(BaseModel):
    firm_type: FirmType
    firm_id: int
    name: str
    website: str | None
    website_source: str | None
    # Flat list; the FE groups by ``provider``.
    values: list[ExtractedValue]

File: frontend/lib/extraction-analytics.ts
// API client for the Extraction Analytics page (/settings/extractions).
// Mirrors backend/app/schemas/extraction_analytics.py.

import { apiRequest, buildApiPath } from "@/lib/api";

export type FirmType = "broker_dealer" | "advisor" | "institutional_investor";
export type Surface = "contact" | "discovered_email" | "website";
export type ValueKind = "email" | "phone" | "linkedin" | "website";

export interface ProviderCount {
  provider: string;
  label: string;
  surface: Surface;
  firm_count: number;
  value_count: number;
}

export interface ProviderSummaryTotals {
  total_contacts_attributed: number;
  total_contacts_unattributed: number;
  total_discovered_emails: number;
  total_enriched_emails: number;
  total_firms_with_website: number;
}

export interface ProviderSummaryResponse {
  contacts: ProviderCount[];
  discovered_emails: ProviderCount[];
  websites: ProviderCount[];
  totals: ProviderSummaryTotals;
}

export interface FirmProviderCount {
  provider: string;
  label: string;
  count: number;
}

export interface FirmExtractionRow {
  firm_type: FirmType;
  firm_id: number;
  name: string;
  crd_number: string | null;
  website_source: string | null;
  contact_total: number;
  discovered_email_total: number;
  provider_counts: FirmProviderCount[];
  last_enriched_at: string | null;
}

export interface FirmExtractionListResponse {
  items: FirmExtractionRow[];
  total: number;
}

export interface ExtractedValue {
  kind: ValueKind;
  value: string;
  provider: string;
  label: string;
  contact_id: number | null;
  contact_name: string | null;
  contact_title: string | null;
  confidence: number | null;
  source_table: "contact" | "discovered_email" | "firm";
}

export interface FirmExtractionDetailResponse {
  firm_type: FirmType;
  firm_id: number;
  name: string;
  website: string | null;
  website_source: string | null;
  values: ExtractedValue[];
}

export async function getExtractionSummary(): Promise<ProviderSummaryResponse> {
  return apiRequest<ProviderSummaryResponse>(
    "/api/v1/extraction-analytics/summary",
  );
}

export interface ListExtractionFirmsParams {
  firm_type?: FirmType;
  provider?: string;
  q?: string;
  page?: number;
  limit?: number;
}

export async function listExtractionFirms(
  params: ListExtractionFirmsParams = {},
): Promise<FirmExtractionListResponse> {
  return apiRequest<FirmExtractionListResponse>(
    buildApiPath("/api/v1/extraction-analytics/firms", {
      firm_type: params.firm_type,
      provider: params.provider,
      q: params.q,
      page: params.page,
      limit: params.limit,
    }),
  );
}

export async function getFirmExtractionValues(
  firmType: FirmType,
  firmId: number,
): Promise<FirmExtractionDetailResponse> {
  return apiRequest<FirmExtractionDetailResponse>(
    `/api/v1/extraction-analytics/firms/${firmType}/${firmId}/values`,
  );
}
